guardian/typerighter
# Add LanguageTool to our pool of matchers, and enable its default rules via the sheet

## What does this change?

Add LanguageTool to our pool of matchers, and enable its default rules via our sheet.

<img width="285" alt="Screenshot 2020-09-21 at 16 03 14" src="https://user-images.githubusercontent.com/7767575/93790621-269f7880-fc2b-11ea-9360-619438869c5b.png">

## How to test

Working locally, or in CODE, take a look in CODE at the palette of LanguageTool rules in the CODE sheet. Enable or disable rules and refresh the rules via UI available at the `/rules`endpoint. Do they behave as expected? ([LanguageTool's live demo](https://languagetool.org/) is a good point of comparison.)

## How can we measure success?

Anyone with access to the relevant sheet can enable or disable LanguageTool rules.

## Have we considered potential risks?

We'll need to revise almost immediately to play nicely with #75. 

## Notes

The rules produced by LanguageTool don't correspond to our usual way of handling categories, which we strongly suspect is due for revision.

With RegexRules, there's a 1-1 mapping between category and matcher. For LT that relationship is broken – there's a 1-1 mapping between a category and matcher, but a many-1 mapping between a _rule_ and a category – so different rules within the same matcher 'category' can produce different categories when matches are made.

This is useful for now, as otherwise we'd see 'Default LanguageTool rules' in our category output for LT matches, which is no use to anybody at present – but it's not obvious to anyone familiar with the way rules are structured.

In future, the Category entity is likely to be replaced with a Tag entity with a many-many relationship, which should solve these sorts of problems.
